docs/core/integration-quality-scale/rules/iqs019.md

```diff
@@ -0,0 +1,107 @@
+---
+title: "IQS019 - Can be discovered"
+---
+
+IQS019 is as follows:
+
+Can be discovered
+
+## Reasoning
+
+A lot of devices have the ability to be discovered.
+This can happen using one of the following methods:
+- mDNS
+- SSDP
+- Bluetooth
+- DHCP
+- USB
+
+This is a great way to make it easier for users to find and set up devices, since they don't have to manually look up which integration to use and then enter the host.
+This greatly reduces the effort required to set up a device and thus improves the user experience.
+
+Using a network based setup, also allows the configuration of the integration to be updated once the device receives a new IP address.
+
+## Example implementation
+
+In the following example, the integration is discoverable using mDNS.
+The device would make itself discoverable by providing a `_mydevice._tcp.local.` service.
+Home Assistant will pick this up and start a discovery flow for the user.
+The user will then be able to confirm the discovery and set up the integration.
+
+`manifest.json`:
+```json
+{
+  "zeroconf": ["_mydevice._tcp.local."]
+}
+```
+
+`config_flow.py`:
+```python
+class MyConfigFlow(ConfigFlow, domain=DOMAIN):
+    """My config flow."""
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self.data: dict[str, Any] = {}
+
+    async def async_step_zeroconf(
+        self, discovery_info: zeroconf.ZeroconfServiceInfo
+    ) -> ConfigFlowResult:
+        """Handle zeroconf discovery."""
+        self.data[CONF_HOST] = host = discovery_info.host
+
+        await self.async_set_unique_id(discovery_info.properties["serialno"])
+        self._abort_if_unique_id_configured(updates={CONF_HOST: host})
+
+        client = MyClient(host)
+        await self.client.get_data()
+
+        return await self.async_step_discovery_confirm()
+
+    async def async_step_discovery_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Confirm discovery."""
+        if user_input is not None:
+            return self.async_create_entry(
+                title="MyIntegration",
+                data={CONF_HOST: self.data[CONF_HOST]},
+            )
+
+        self._set_confirm_only()
+        return self.async_show_form(step_id="discovery_confirm")
+
+    async def async_step_user(
```

Comment:
This body doesn't really do anything. I think either we should remove the user step method or show a plausible user step.

Reply:
I feel like its more of an "oh we set the same unique id here" and that we show that we can set up the device manually as well and that we don't completely rely on discovery

Reply:
Oh wait, the example doesn't even set an unique id. If we would add that, would that be better?

Reply:
Yes, the user step should make sense in the context of the example. If the user step is shown to show a manual step and a discovery step and their difference that's ok.

Reply:
I'd add the same logic for setting unique id and check for getting data in the user step as in the discovery step.